### ISSUE TITLE
Stdout blank running "cmd.exe /c ver"

### ISSUE THREAD

**Issue by IBwWG** [HaxeFoundation/haxe#5430]
On haxe 3.2.1, win7sp1x64.

From
http://stackoverflow.com/a/13214734/16777
and
https://github.com/HaxeFoundation/haxe/issues/4509#issuecomment-136129189
...one would think you could combine the two to capture the version.  However:

```haxe
var output = new sys.io.Process("cmd.exe", ["/c","ver"]).stdout.readAll().toString();
trace(output.length);
```

...gives `0`.

This is also the case for non-built-in-commands, e.g. wmic.exe (when wrapped via `cmd.exe`; if they aren't, you get a `Process creation failure` run-time error.)  Yet batch files containing the same lines, run from haxe, return the correct data into `output`.  Why does stdout not seem to work without a batch file as a middleman?

**Comment by Simn** [comment]
I think you have to call `exitCode()` on the Process instance first.

**Comment by IBwWG** [comment]
Oh, sorry, even with that, same result.  (I had simplified the code a bit.)

**Comment by IBwWG** [comment]
Interesting, if I trace stderr I get:
```
:,'"ver' is not recognized as an internal or external command,

operable program or batch file.
```

That extra quote there may indicate a problem...

**Comment by IBwWG** [comment]
Come to think of it, I had to stop using `args` in another context too because it was giving a different result than putting everything into the command name instead.  Maybe this is the cause of that too...

**Comment by IBwWG** [comment]
However, I can't do that in this case, because of this process creation error:

```
Process creation failure : cmd.exe /c ver
Process creation failure : cmd.exe /c wmic OS get TotalVisibleMemorySize /Value
```

**Comment by IBwWG** [comment]
According to http://api.haxe.org/sys/io/Process.html that approach should work, given that I can type `cmd.exe /c ver` on the command line and it works (without any quoting.)

**Comment by theinternetftw** [comment]
There's a problem with the quote escaping in hxcpp, best I can tell.  You can get around it by passing the arguments in the command itself.  That's not working for you because you're doing this ```new Process("cmd.exe /c ver", []);``` and the empty array doesn't cut it to trigger the alternative behavior, you have to use null (```new Process("cmd.exe /c ver", null);```, that is).

**Comment by theinternetftw** [comment]
After looking at it again, the quote problem is that hxcpp quotes every argument no matter what, which messes up cmd.exe's option handling.

You can tell this because the following fails in the exact same manner when typed into a cmd prompt:

```"cmd.exe" "/c" "ver"```

[This](https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/) seems to be the bible for getting this kind of thing right on windows.

**Comment by andyli** [comment]
It is probably fixed in Haxe 3.3. See https://github.com/HaxeFoundation/haxe/wiki/Breaking-changes-in-Haxe-3.3.0#syscommand-and-sysioprocess

**Comment by theinternetftw** [comment]
Just tested in 3.3.  Error message remains the same.

And if you just look at the code for hxcpp ([3.3 Version](https://github.com/HaxeFoundation/hxcpp/blob/master/src/hx/libs/std/Process.cpp#L163), [3.2 Version](https://github.com/HaxeFoundation/hxcpp/blob/master/project/libs/std/Process.cpp#L123)), you can see that every arg is quoted regardless.  

(That is, if I'm reading hxcpp's source right and those two files are for the versions I think they are.  I don't even use haxe, I just filed a bug for something that does and started having fun \:D)

**Comment by andyli** [comment]
I confirm that the Windows `cmd.exe` has to be called without using the `args` of the Haxe `Sys.command` / `Process` API. i.e. the following code will work as expected:
```haxe
class Test {
    static function main() {
        var p = new sys.io.Process("cmd.exe /c ver"); // or new sys.io.Process("cmd.exe /c ver", null);
        trace(p.stdout.readAll());
    }
}
```

This is because cmd.exe does not use the standard C/C++ command line argument parsing rule, thus it doesn't work with how we escape the arguments. Read more: http://stackoverflow.com/questions/34448131/does-the-windows-cmd-exe-parse-arguments-differently

**Comment by IBwWG** [comment]
@andyli but doesn't this give you a `Process creation failure` like it does for me on haxe 3.2.1?

**Comment by IBwWG** [comment]
I put this into Test.hx (just what you had, plus the import line needed to compile it):

```haxe
import sys.io.Process;
class Test {
    static function main() {
        var p = new sys.io.Process("cmd.exe /c ver", null); // or new sys.io.Process("cmd.exe /c ver", null);
        trace(p.stdout.readAll());
    }
}
```

Then I ran `haxe -main Test -cpp .\` followed by `Test.exe`.  The result was:

```
Error : Process creation failure : cmd.exe /c ver
```

This is on haxe 3.2.1 on win7sp1x64.  Doesn't that mean this is a bug / should stay open?

**Comment by andyli** [comment]
It's fixed in Haxe 3.3. I've tested the code on Windows for both the Neko and C++ targets.
Generally once we've moved forward to a new version (3.3), we do not release bug fix to the old one (3.2). So the fix will be in 3.3 and the issue is closed.
